scriptworker/utils.py

```diff
@@ -355,19 +354,40 @@ def create_temp_creds(client_id, access_token, start=None, expires=None,
     if tasks:
         await asyncio.wait(tasks)
         for task in tasks:
-            exc = task.exception()
-            if exc is None:
-                succeeded_results.append(task.result())
-            else:
+            exc = get_future_exception(task)
+            if exc:
```

Comment:
Given that the new function returns both `False` and `None` with different meanings, I think we should handle both cases explicitly here. From the comment below, it sounds like you don't expect to get `False` in this context, it might be reasonable to just log in that case, although maybe bailing out is better if we hit an impossible situation?

Reply:
We've already waited on all the tasks above, which I believe means they've all completed or thrown exceptions. Since `False` means the future hasn't finished yet, I don't expect to hit that, and adding another if block is another thing I need to cover in tests -- I suppose we could reach that block by mocking enough things, but at the time of writing I thought it would make more sense to leave it off.

Given the above, do you still need to check for `False` ?

Reply:
I have two thoughts on this (which are really two different ways of looking at the same thing:
1. As a reader, when I look at this code, I see the `get_future_exception`, see that it has three possible return values and then am wondering why no result result is being treated as success and wondering if it was mistake because two of the return values happen to be false-y (looking at it through a rust lens, I'd expect there two be three case in handling the return value). Given that the you expect the `False` case never to happen, I'd be fine with that branch being marked `pragma: nocover`.
2. Looking at `get_future_exception`, it looks like a fairly trivial wrapper around `future.exception()`. It looks like it exists to turn a `InvalidStateError` into a `False` return value. As a reader, I expect that there is going to be somewhere where this case is not exception and you want to handle all three case in parallel ways. But when I go look at the only caller, it appears that you don't actually handle the `False` case because it is shouldn't ever happen. That suggests that just using `future.exception()` would be better here, and let the impossible exception bubble up, rather than be silently treated as a success.

Reply:
We also use `get_future_exception` in `run_task`: `timeout_exc = get_future_exception(timeout_task)`. I can certainly use `future.exception()` here, and only use `get_future_exception` in the one place.